src/common.cc

```diff
@@ -111,4 +121,337 @@ Join(const std::vector<std::string>& vec, const std::string& delim)
+/// returns length in host byte order.
+///
+/// \param[in] length
+/// \param[in] host_byte_order whether original length is in host byte order
+/// \param[out] host_length length in host byte order
+/// \return length with byte order swapped
+///
+uint32_t
+bswap_length(
+    const uint32_t length, const bool host_byte_order, uint32_t* host_length)
+{
+  uint32_t swapped = bswap(length);
+  *host_length = host_byte_order ? length : swapped;
+  return swapped;
+}
+
+/// Swaps length byte order in place with support for non coniguous
+/// arrays and returns length in host byte order. For
+/// non-contiguous arrays the initial call must pass an empty partial result.
+///
+/// \param[in, out ] base pointer to array with length
+/// \param[in]] byte_size size of array
+/// \param[in] host_byte_order whether original length is in host byte order
+/// \param[out] host_length length in host byte order
+/// \param[in, out] partial_result vector for partial results.
+///
+void
+bswap_length_partial(
+    char* base, size_t byte_size, const bool host_byte_order,
+    uint32_t* host_length, std::vector<char*>& partial_result)
+{
+  *host_length = 0;
+
+  if (partial_result.size()) {
+    for (size_t offset = 0,
+                remaining = sizeof(uint32_t) - partial_result.size();
+         (remaining > 0) && (offset < byte_size);
+         ++offset, ++base, --remaining) {
+      partial_result.emplace_back(base);
+    }
+    if (partial_result.size() == sizeof(uint32_t)) {
+      char* host_length_buffer = reinterpret_cast<char*>(host_length);
+      for (size_t left_index = 0, right_index = sizeof(uint32_t) - 1;
+           left_index < right_index; ++left_index, --right_index) {
+        char temp = *partial_result[right_index];
+        *partial_result[right_index] = *partial_result[left_index];
+        *partial_result[left_index] = temp;
+        if (host_byte_order) {
+          host_length_buffer[right_index] = *partial_result[left_index];
+          host_length_buffer[left_index] = *partial_result[right_index];
+        } else {
+          host_length_buffer[right_index] = *partial_result[right_index];
+          host_length_buffer[left_index] = *partial_result[left_index];
+        }
+      }
+      partial_result.clear();
+      return;
+    }
+    return;
+  }
+
+  if (byte_size >= sizeof(uint32_t)) {
+    uint32_t* value = reinterpret_cast<uint32_t*>(base);
+    *value = bswap_length(*value, host_byte_order, host_length);
+    return;
+  }
+
+  for (; byte_size; --byte_size, ++base) {
+    partial_result.emplace_back(base);
+  }
+}
+
+/// Swaps byte order in place
+///
+/// \param[in, out] base pointer to array of data type elements to convert
+/// \param[in] byte_size size of array in bytes
+///
+template <typename T>
+void
+SwapEndian(char* base, size_t byte_size)
+{
+  T* value = reinterpret_cast<T*>(base);
+  for (size_t offset = 0; offset < byte_size; offset += sizeof(T), ++value) {
+    *value = bswap(*value);
```

Comment:
IMO we should look into using SIMD for these functions as they will be on the critical inference path. AVX2 shipped in 2013 which would be nice to target. PowerPC looks like it has [support for SIMD](https://stackoverflow.com/questions/66195879/simd-extensions-on-power-compiler-flags-and-processor-support) as well.

Reply:
As this code will only be compiled on architectures where the native byte order is big endian - do you think that is something we could add as a follow on optimization? My thinking is to provide initial functionality / get feedback on the initial approach.

Reply:
I like Kyle's perf recommendations, but I think they could be medium/high effort for low impact given the rarity at which we encounter big endian machines. I would advocate leaving it as a follow-up and waiting for reports of slowness.

Reply:
as discussed offline - we'll revisit the performance optimizations in a later PR